zephyr/wrapper.c

```diff
@@ -242,18 +226,54 @@ void *rzalloc(enum mem_zone zone, uint32_t flags, uint32_t caps, size_t bytes)
 void *rballoc_align(uint32_t flags, uint32_t caps, size_t bytes,
 		    uint32_t alignment)
 {
-	return heap_alloc_aligned_cached(&sof_heap, alignment, bytes);
+	void *ptr;
+	struct sof_multiheap_params params;
+
+	params.flags = flags;
+	params.caps = caps;
+	params.zone = SOF_MEM_ZONE_BUFFER;
+
+	k_spinlock_key_t key = k_spin_lock(&multi_heap_lock);
+
+	ptr = sys_multi_heap_aligned_alloc(&multi_heap, &params, alignment, bytes);
+	k_spin_unlock(&multi_heap_lock, key);
+
+	return ptr;
 }
 
 /*
  * Free's memory allocated by above alloc calls.
  */
 void rfree(void *ptr)
 {
+	struct sys_heap_ext *heap;
+	void *ptr_uncached = ptr;
 	if (!ptr)
 		return;
 
-	heap_free(&sof_heap, ptr);
+#ifdef CONFIG_SOF_ZEPHYR_HEAP_CACHED
+	if (is_cached(ptr))
+		ptr_uncached = z_soc_uncached_ptr(ptr);
+#endif
+
+	/* get heap and metadata from the multiheap, always provide uncached alias */
+	k_spinlock_key_t key = k_spin_lock(&multi_heap_lock);
+
+	heap = sys_multi_heap_get_heap(&multi_heap, ptr_uncached);
+
+#ifdef CONFIG_SOF_ZEPHYR_HEAP_CACHED
+	uint32_t heap_caps = *(uint32_t *)(heap->metadata);
+
+	if (heap_caps & SOF_MEM_CAPS_CACHE)
+	/* if the memory region has cache capabilities, invalidate the cache, no writeback */
+		z_xtensa_cache_flush_inv(ptr, sys_heap_usable_size(heap, ptr_uncached));
+#endif
+
+	/* release memory block */
+	sys_heap_free(heap->heap, ptr_uncached);
+	k_spin_unlock(&multi_heap_lock, key);
+
+	/* TODO: do other operations, like powering off memory banks, unmapping TLB etc. */
```

Comment:
Can we move the allocator APIs to a src/lib/zephyr_alloc.c. This will help reduce zephyr/wrapper.c which we really need to delete once we are fully aligned at the RTOS API level.